### PR TITLE
Feature/Lookit - Update Account Info, Demographic Data, Add/Edit Children, Update Email Preferences [LEI-431]

### PR DESCRIPTION
**Pull request by pattisdr**
# Ticket
https://openscience.atlassian.net/browse/LEI-431

# Purpose

Lookit-side, allow participants to edit their account information, demographic data, password, add/edit children, and update email preferences.

Also addresses CR from https://github.com/CenterForOpenScience/lookit-api/pull/17

Modifies some of the API querysets w/ distinct().

# Changes 
## Edit Account Info
screencapture-localhost-8000-account-1500394610618

## Edit Demographics
screencapture-localhost-8000-account-demographics-1500394673796

## View Children
screencapture-localhost-8000-account-children-1500394846403

## Add Child
screencapture-localhost-8000-account-children-1500394858441

## Update/Delete Child
screencapture-localhost-8000-account-children-9e1a7762-9673-4ec0-8547-599034c4c0b0-1500394876601

## Update Email Preferences
screencapture-127-0-0-1-8000-account-email-1500414197568

# TODO
- [x] Account Information Page
- [x] Demographic Survey Page
- [x] Children Information (Add/Edit Children)
- [x]  Email Preferences

### CODE REVIEW

File: web/templates/web/demographic-data-update.html [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  {% endif %}
+{% endblock %}
+{% block content %}
+    <div class='lookit-row lookit-page-title'>
+        <h2 class='container'>My Account</h2>
+    </div>
+    <div class="container">
+        <div class="row account-edit">
+            <div class="col-md-4 mb-lg">
+                <a class="btn btn-default btn-md btn-block" href="/account"><strong>Account information</strong><br> Change your name, email, or password.</a>
```
Comment (cwisecarver) [inline web/templates/web/demographic-data-update.html]: You could use that partial here.

Alternatively, you could make a "layout" template that this page (and the others like it) could extend, like it currently extends web/base.html and put all your global boilerplate in that. That template could extend web/base.html so you don't have to duplicate it's contents. Then you create your own block in your layout that's inside block content.

File: web/templates/web/participant-email-preferences.html [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  {% endif %}
+{% endblock %}
+{% block content %}
+    <div class='lookit-row lookit-page-title'>
+        <h2 class='container'>My Account</h2>
+    </div>
+    <div class="container">
+        <div class="row account-edit">
+            <div class="col-md-4 mb-lg">
+                <a class="btn btn-default btn-md btn-block" href="/account"><strong>Account information</strong><br> Change your name, email, or password.</a>
```
Comment (cwisecarver) [inline web/templates/web/participant-email-preferences.html]: Url template tags or layout or partial.

File: web/templates/web/participant-update.html [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  {% endif %}
+{% endblock %}
+{% block content %}
+    <div class='lookit-row lookit-page-title'>
+        <h2 class='container'>My Account</h2>
+    </div>
+    <div class="container">
+        <div class="row account-edit">
+            <div class="col-md-4 mb-lg ">
+                <a class="btn btn-default btn-md btn-block active" href="/account"><strong>Account information</strong><br> Change your name, email, or password.</a>
```
Comment (cwisecarver) [inline web/templates/web/participant-update.html]: Same as others

File: web/views.py [hunk 4]
```diff
@@ -84,6 +86,81 @@ def form_valid(self, form):
...
 
+class ChildrenListView(generic.CreateView):
+    """
+    Allows user to view a list of current children and add children
+    """
+    template_name = 'web/children-list.html'
+    model = Child
+    form_class = forms.ChildForm
+
+    def dispatch(self, request, *args, **kwargs):
```
Comment (cwisecarver) [inline web/views.py]: I literally just did this wrong in my PR that hasn't been merged yet ... I'll go fix it. But the right way is to inherit from [this](https://ccbv.co.uk/projects/Django/1.11/django.contrib.auth.mixins/LoginRequiredMixin/) and CreateView.

It'll force them to login if they're not logged in.

File: web/views.py [hunk 5]
```diff
@@ -84,6 +86,81 @@ def form_valid(self, form):
...
+
+class ChildUpdateView(generic.UpdateView):
+    """
+    Allows user to view a list of current children and add children
+    """
+    template_name = 'web/child-update.html'
+    model = Child
+    form_class = forms.ChildForm
+
+    def dispatch(self, request, *args, **kwargs):
```
Comment (cwisecarver) [inline web/views.py]: Same here with the `LoginRequiredMixin`

File: web/views.py [hunk 6]
```diff
@@ -93,6 +170,11 @@ class DemographicDataCreateView(generic.CreateView):
     model = DemographicData
     form_class = forms.DemographicDataForm
 
+    def dispatch(self, request, *args, **kwargs):
```
Comment (cwisecarver) [inline web/views.py]: LoginRequiredMixin, this should probably end up on almost all our views that aren't the flat pages of lookit or login/signup

File: web/views.py [hunk 7]
```diff
@@ -101,3 +183,114 @@ def form_valid(self, form):
...
     def get_success_url(self):
         return reverse('web:studies-list')
+
+class DemographicDataUpdateView(DemographicDataCreateView):
+    """
+    Allows user to update demographic data - but actually creates new version instead of updating old one.
+    """
+    template_name = 'web/demographic-data-update.html'
+
+    def dispatch(self, request, *args, **kwargs):
```
Comment (cwisecarver) [inline web/views.py]: LoginRequiredMixin

File: web/views.py [hunk 8]
```diff
@@ -101,3 +183,114 @@ def form_valid(self, form):
...
+    def get_success_url(self):
+        return reverse('web:demographic-data-update')
+
+    def get_initial(self):
+        """
+        Returns the initial data to use for forms on this view.
+        """
+        demographic_data = self.request.user.latest_demographics or None
+        if demographic_data:
+            demographic_data = demographic_data.__dict__
```
Comment (cwisecarver) [inline web/views.py]: Shadowing this variable is creepy. Maybe demographic_data_dict?

File: web/views.py [hunk 9]
```diff
@@ -101,3 +183,114 @@ def form_valid(self, form):
...
+            demographic_data.pop('id')
+            demographic_data.pop('uuid')
```
Comment (cwisecarver) [inline web/views.py]: We should set the `demographic_data_dict['previous_id']` to `demographic_data.id`.

File: web/views.py [hunk 10]
```diff
@@ -101,3 +183,114 @@ def form_valid(self, form):
...
+
+class ParticipantEmailPreferencesView(generic.UpdateView):
+    """
+    Allows a participant to update their email preferences - when they can be contacted.
+    """
+    template_name = 'web/participant-email-preferences.html'
+    model = User
+    form_class = forms.EmailPreferencesForm
+
+    def dispatch(self, request, *args, **kwargs):
```
Comment (cwisecarver) [inline web/views.py]: LoginRequiredMixin

File: web/views.py [hunk 11]
```diff
@@ -101,3 +183,114 @@ def form_valid(self, form):
...
+class ParticipantUpdateView(generic.UpdateView):
+    """
+    Allows a participant to update their names and password
+    """
+    template_name = 'web/participant-update.html'
+    model = User
+    form_class = forms.ParticipantUpdateForm
+    second_form_class = forms.ParticipantPasswordForm
+
+    def dispatch(self, request, *args, **kwargs):
```
Comment (cwisecarver) [inline web/views.py]: LoginRequiredMixin

File: web/templates/web/children-list.html [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+           });
+        })
+    </script>
+    <div class='lookit-row lookit-page-title'>
+        <h2 class='container'>My Account</h2>
+    </div>
+    <div class="container">
+        <div class="row account-edit">
+            <div class="col-md-4 mb-lg">
+                <a class="btn btn-default btn-md btn-block" href="/account"><strong>Account information</strong><br> Change your name, email, or password.</a>
```
Comment (pattisdr) [inline web/templates/web/children-list.html]: ehh. i was being lazy! moved it to partial template.

**Comment by pattisdr** [comment]
All changes addressed @cwisecarver

**Comment by cwisecarver** [comment]
@pattisdr I'm going to go ahead and merge it just to get it done. If you wouldn't mind changing it in whatever PR you're working on now, standard practice for "partial" templates is to prefix them with an underscore so that they're easy to distinguish from regular ones. Just toss an _ in front of accounts-navigation.html please. Thanks!

**Pull request merged**